### PR TITLE
Events migration

### PR DESCRIPTION
**Pull request by adam-james-v**
Part of Epic #20289 

# Events and Timelines Migrations
This PR adds the basic machinery for *Events* and *Timelines*.

The relevant product doc is [Allow users to annotate (and use) meaningful events](https://www.notion.so/metabase/Allow-users-to-annotate-and-use-meaningful-events-b06ba695b07541d09fc3ac23ea1852b8) 

## API Endpoints (So far)

***timeline***
- [x] POST /api/timeline
- [x] GET /api/timeline/:id
- [x] GET /api/timeline/:id?include=events
- [x] GET /api/timeline/:id?include=events&archived=true
- [x] PUT /api/timeline/:id
- [x] DELETE /api/timeline/:id 

***timeline-event***
- [x] POST /api/timeline-event
- [x] GET /api/timeline-event/:id
- [x] PUT /api/timeline-event/:id
- [x] DELETE /api/timeline-event/:id 

***timelines via collection***
- [x] GET /api/collection/root/timelines
- [x] GET /api/collection/root/timelines?include=events
- [x] GET /api/collection/root/timelines?include=events&archived=true
- [x] GET /api/collection/:id/timelines
- [x] GET /api/collection/:id/timelines?include=events
- [x] GET /api/collection/:id/timelines?include=events&archived=true

***timelines via card***
- [x] GET /api/card/:id/timelines
- [x] GET /api/card/:id/timelines?include=events
- [x] GET /api/card/:id/timelines?include=events&archived=true

***Endpoint for Getting Events within a Range***
- [x] GET /api/card/:id/timelines?start=X&end=Y

### CODE REVIEW

File: src/metabase/api/timeline.clj [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      (= include "events")
+      (timeline-event/include-events-singular {:events/all? archived?}))))
+
+(api/defendpoint PUT "/:id"
+  "Update the [[Timeline]] with `id`. Returns the timeline without events. Archiving a timeline will archive all of the
+  events in that timeline."
+  [id :as {{:keys [name description icon collection_id archived] :as timeline-updates} :body}]
+  {name          (s/maybe su/NonBlankString)
+   description   (s/maybe s/Str)
+   ;; todo: there are six valid ones. What are they?
```
Comment (escherize) [inline src/metabase/api/timeline.clj:55]: Is this todo still relevant? (I found a photo of them, but don't know what they're called.) https://www.figma.com/file/AOcPjJRDDwUGYB6PvuFfVh/Events?node-id=278%3A28252

File: src/metabase/api/timeline.clj [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+   icon          (s/maybe s/Str)
+   collection_id (s/maybe su/IntGreaterThanZero)
+   archived      (s/maybe s/Bool)}
+  ;; todo: icon is valid
```
Comment (escherize) [inline src/metabase/api/timeline.clj:59]: I think you will need to make the icon schema an enum.

File: src/metabase/api/timeline.clj [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+(api/defendpoint DELETE "/:id"
+  "Delete a [[Timeline]]. Will cascade delete its events as well."
+  [id]
+  (api/write-check Timeline id)
+  (db/delete! Timeline :id id)
+  api/generic-204-no-content)
+
+
+;; todo: icons
+;; todo: how does updated-at work?
```
Comment (escherize) [inline src/metabase/api/timeline.clj:80]: more todos to delete or do

File: src/metabase/api/timeline_event.clj [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            [metabase.util.schema :as su]
+            [schema.core :as s]
+            [toucan.db :as db]))
+
+(api/defendpoint POST "/"
+  "Create a new [[TimelineEvent]]."
+  [:as {{:keys [name description timestamp time_matters timezone icon timeline_id archived] :as body} :body}]
+  {name         su/NonBlankString
+   description  (s/maybe s/Str)
+   ;; todo: find schema for timestamps?
```
Comment (escherize) [inline src/metabase/api/timeline_event.clj:20]: I didn't see one in the code base, I could have missed it or you might need to make one.

**Review by escherize — CHANGES_REQUESTED** [review]
Looks good! Just a few todos that might need some attention.

File: src/metabase/api/timeline.clj (hunk 1 above)
Comment (dpsutton) [reply src/metabase/api/timeline.clj:55]: we got the acceptable icons pretty late. I think its fine for followup work. We're trying to get this into master so our FE work can get unblocked.

File: src/metabase/api/timeline_event.clj (hunk 4 above)
Comment (dpsutton) [reply src/metabase/api/timeline_event.clj:20]: I ended up making one. I guess we forgot to put it here. Good call

Comment (escherize) [reply src/metabase/api/timeline_event.clj:20]: Cool, sounds like hooking up the timestamp schema is all that's needed now. I'll appove

**Review by escherize — APPROVED** [review]

**Pull request merged**